README.md
# Securing endpoints in Express with OAuth JWT tokens
       
[![Quality](https://img.shields.io/badge/quality-test-yellow)](https://curity.io/resources/code-examples/status/)
[![Availability](https://img.shields.io/badge/availability-source-blue)](https://curity.io/resources/code-examples/status/)


This library allows you to secure your Express endpoints with JWTs. The implementation uses a JWKS endpoint of an
Authorization Server to get keys required for the verification of the token signature.

## Running the example

1. Download the code and install dependencies running `npm i`.
2. In `settings.js` in the `example` directory set the URI of the JWKS endpoint exposed by the Authorization Server.
3. Start the demo server with `npm run example`.
4. The server exposes endpoints `/secured/token`, `/secured/scope` and `/secured/claim` which present different options
of securing the endpoints.

## Installing the dependency

Use `npm` to install the library in your project:

```bash
npm install express-oauth-jwt
```

## Usage

### Preparing the Middleware

The `secure` middleware needs a method to obtain keys. You can use the `jose`-provided method, or create your own instance.
The method is responsible for obtaining signature verification keys for a concrete token, and has the signature:

`function(headerParameters: JWSHeaderParameters, input: FlattenedJWSInput): Promise<KeyLike | Uint8Array> | KeyLike | Uint8Array`

The easiest way is to use the method provided by the `jose` library. The method caches the result of the JWKS endpoint request
in memory to limit sending requests to the server.

```javascript
const jwksService = createRemoteJWKSet(new URL(settings.jwks_uri))
const midleware = secure(jwksService)

```

### Securing an endpoint

To secure an endpoint apply the `secure` middleware to it. Any endpoint with this middleware applied will require a valid
JWT token sent in the `Authorization` header in the form `Bearer <token_value>`. The token, if valid, will be decoded
and all its claims will be set in the request in a `claims` field.

If the token does not pass validation, a 403 response will be returned. If no JWT token is found in the request, a 401
response will be returned. The response will have the `WWW-Authenticate` header set with detailed information on
the error (as specified in the [RFC 6750: The OAuth 2.0 Authorization Framework: Bearer Token Usage](https://tools.ietf.org/html/rfc6750)).

If you want the `WWW-Authenticate` to return a `realm` value, pass the middleware an `options` object with the `realm`
option set.

```javascript
const { secure } = require('express-oauth-jwt');
router.use(secure(jwksService, { realm: 'my-realm' }));
```

The `WWW-Authenticate` header in the error response will then might look like this:

```curl
WWW-Authenticate: Bearer realm="my-realm", error="invalid_token", error_description="..."
```

### Authorizing the request based on scopes

In order to limit the request to given scope values pass the middleware an `options` object with a list of strings in the
`scope` field:

```javascript
const secure = require('express-oauth-jwt');
router.use(secure(jwksService, { scope: ["scope1", "scope2"] }));
```

### Authorizing the request based on claims

In order to limit the request based on the presence or value of concrete claims, pass the middleware an `options` object
with a list of claims in the `claims` field. Each claim should be an object with at least the `name` field. Optionally
you can set the `value` field. If only the `name` field is set then the validator checks whether the claim
exists in the token. If `value` is set as well, then the value of the claim in the token must also match.

```javascript
const secure = require('express-oauth-jwt');
router.use(secure(jwksService, { claims: [ { name: "someClaim", value: "withValue" } ] }));
```

## Using Opaque tokens

This middleware uses JSON Web Tokens, which can be easily decoded offline into a JSON object containing claims about
the user or client sending the request. But this is not always the case that JWT tokens are used for authorization.
Sometimes the Authorization Server will issue an opaque token, which is just an identifier of the user's data, and the
data itself is kept safely with the Authorization Server.

If you use opaque tokens for authorization then the token needs to be exchanged for data associated with it. It
cannot be decoded, as the value of the token does not contain any data. In such situation, the best approach is to use the
Phantom token approach - let your API gateway exchange the opaque token for a JWT. This way your service will always
receive a JWT which can be decoded with the approach shown here.

You can find out more on the Phantom token approach in the
[Phantom Token Pattern](https://curity.io/resources/architect/api-security/phantom-token-pattern/) article.

If you're using Curity Identity Server you can learn how to enable Phantom tokens with the help of
[this tutorial](https://curity.io/resources/operate/tutorials/integration/introspect-with-phantom-token/).

## Access token claims in Request object

The `secure` middleware adds claims obtained from the access token into the Express request object, in the `claims`
field. You can access these claims in any other middleware which is next in chain and in the controllers.

```javascript
function getLibraryData(req, res) {
    if (req.claims.myClaim == 'someValue') {
        ...
    }
    ...
}
```

## Questions and Support

For questions and support, contact Curity AB:

> Curity AB
>
> info@curity.io
> https://curity.io

Copyright (C) 2020 Curity AB.


CHANGELOG.md
# Changelog

## Version 2.0.2
Released 2023-07-26

- Authorizing the request based on claim values now properly handles array claims. The request is authorized if the required value is one of the elements of the array.

## Version 2.0.1
Released 2022-10-07

- Updated jose library to 4.10.0

## Version 2.0.0
Released: 2022-04-13

- Updated jose library to version 4.
- Removed `jwksService` class, as the functionality of downloading keys from a JWKS endpoint is now provided by jose directly.
- The `secure` middleware now needs a `getKey` function, not a `jwksService` as a constructor parameter.

## Version 1.1.0
Released: 2021-04-02

## Version 1.0.1
Released: 2020-04-29

## Version 1.0.0
Released: 2020-04-29

Initial version of the library.


lib/secure.js
/*
 *  Copyright 2020 Curity AB
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

const { jwtVerify } = require('jose');
const { InsufficientScopeError, TokenClaimsError } = require('./errors');

module.exports = function(getKey, options) {
    const resolvedOptions = {
      scope: [],
      claims: [],
      realm: '',
      ...options
    };

    const wwwAuthenticate = 'WWW-Authenticate';
    const wwwAuthValue = resolvedOptions.realm !== '' ? 'Bearer realm="' + resolvedOptions.realm + '"' : 'Bearer';

    return async function(req, res, next) {
        // Check if Authorization header is present
        const authHeader = req.headers.authorization;
        if (!authHeader) {
            return res.status(401).append(wwwAuthenticate, wwwAuthValue).send(); 
        }

        // Authorization header should have the form "Bearer <token_value>"
        const headerParts = authHeader.split(' ');

        if (headerParts.length !== 2 || headerParts[0] !== 'Bearer') {
            return res.status(401).append(wwwAuthenticate, wwwAuthValue).send();
        }

        const token = headerParts[1];

        if (!token) {
            return sendErrorResponse(res, 'invalid_token', 'Bearer token is empty');
        }

        return jwtVerify(token, getKey)
            .then((verifiedToken) => {
                if (!tokenHasAllExpectedScopes(verifiedToken.payload.scope, resolvedOptions.scope)) {
                    throw new InsufficientScopeError()
                }

                if (!tokenHasRequiredClaims(verifiedToken.payload, resolvedOptions.claims)) {
                    throw new TokenClaimsError()
                }

                return verifiedToken
            })
            .then((verifiedToken) => {
                req.claims = verifiedToken.payload;
                next();
            })
            .catch(error => {
                if (error instanceof InsufficientScopeError) {
                    return sendErrorResponse(res, 'insufficient_scope', error.message);
                }

                return sendErrorResponse(res, 'invalid_token', error.message.replace(/"/g, "'"));
            });
    }

    function tokenHasAllExpectedScopes(scope, expectedScopes) {
        if (!expectedScopes.length) {
            return true;
        }

        if (!scope || !scope.length) {
            return false;
        }

        for (let index = 0; index < expectedScopes.length; index++) {
            if (!scope.includes(expectedScopes[index])) {
                return false;
            }
        }

        return true;
    }

    function tokenHasRequiredClaims(token, expectedClaims) {
        for (let index = 0; index < expectedClaims.length; index++) {
            const expectedClaim = expectedClaims[index];
            const claim = token[expectedClaim.name];

            if (claim === undefined || claim === null) {
                return false;
            }

            if (Array.isArray(claim)) {
                
                // For array claims such as 'aud', any matching element passes
                return claim.indexOf(expectedClaim.value) !== -1;

            } else {

                // For individual claims such as 'iss', require an exact match
                if (expectedClaim.value && claim !== expectedClaim.value) {
                    return false;
                }
            }
        }

        return true;
    }

    function sendErrorResponse(res, error, description) {
        return res
            .status(403)
            .append(wwwAuthenticate, wwwAuthValue)
            .append(wwwAuthenticate, 'error="' + error + '"')
            .append(wwwAuthenticate, 'error_description="' + description + '"')
            .send();
    }
}
